src/sgame/components/SpikerComponent.cpp

```diff
@@ -19,175 +39,215 @@ void SpikerComponent::HandleDamage(float amount, gentity_t *source, Util::option
+	float missileSize = (float)ma->size;
+	trace_t trace;
+	vec3_t mins, maxs;
+	Vec3 end = Vec3::Load(entity.oldEnt->s.origin) + (SPIKE_RANGE * direction);
+
+	// Test once with normal and once with inflated missile bounding box.
+	for (float traceSize : {missileSize, missileSize * SAFETY_TRACE_INFLATION}) {
+		mins[0] = mins[1] = mins[2] = -traceSize;
+		maxs[0] = maxs[1] = maxs[2] =  traceSize;
+		trap_Trace(&trace, entity.oldEnt->s.origin, mins, maxs, end.Data(), entity.oldEnt->s.number,
+			ma->clipmask, 0);
+		gentity_t* hit = &g_entities[trace.entityNum];
+
+		if (hit && G_OnSameTeam(entity.oldEnt, hit)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
 bool SpikerComponent::Fire() {
 	gentity_t *self = entity.oldEnt;
 	// Check if still resting.
 	if (restUntil > level.time) {
+		logger.Verbose("Spiker #%i wanted to fire but wasn't ready.", entity.oldEnt->s.number);
+
 		return false;
+	} else {
+		logger.Verbose("Spiker #%i is firing!", entity.oldEnt->s.number);
 	}
 
 	// Play shooting animation.
 	G_SetBuildableAnim(self, BANIM_ATTACK1, false);
+	GetBuildableComponent().ProtectAnimation(5000);
 
 	// TODO: Add a particle effect.
 	//G_AddEvent(self, EV_ALIEN_SPIKER, DirToByte(self->s.origin2));
 
 	// Calculate total perimeter of all spike rows to allow for a more even spike distribution.
 	float totalPerimeter = 0.0f;
 
-	for (int row = 0; row < SPIKER_MISSILEROWS; row++) {
-		float altitude = (((float)row + SPIKER_ROWOFFSET) * M_PI_2) / (float)SPIKER_MISSILEROWS;
+	for (int row = 0; row < MISSILEROWS; row++) {
+		float altitude = (((float)row + ROWOFFSET) * M_PI_2) / (float)MISSILEROWS;
 		float perimeter = 2.0f * M_PI * cos(altitude);
 
 		totalPerimeter += perimeter;
 	}
 
 	// Distribute and launch missiles.
+	// TODO: Use new vector library.
 	vec3_t dir, rowBase, zenith, rotAxis;
 
 	VectorCopy(self->s.origin2, zenith);
 	PerpendicularVector(rotAxis, zenith);
 
-	for (int row = 0; row < SPIKER_MISSILEROWS; row++) {
-		float altitude = (((float)row + SPIKER_ROWOFFSET) * M_PI_2) / (float)SPIKER_MISSILEROWS;
+	for (int row = 0; row < MISSILEROWS; row++) {
+		float altitude = (((float)row + ROWOFFSET) * M_PI_2) / (float)MISSILEROWS;
 		float perimeter = 2.0f * M_PI * cos(altitude);
 
 		RotatePointAroundVector(rowBase, rotAxis, zenith, RAD2DEG(M_PI_2 - altitude));
 
 		// Attempt to distribute spikes with equal distance on all rows.
-		int spikes = (int)round(((float)SPIKER_MISSILES * perimeter) / totalPerimeter);
+		int spikes = (int)round(((float)MISSILES * perimeter) / totalPerimeter);
 
 		for (int spike = 0; spike < spikes; spike++) {
 			float azimuth = 2.0f * M_PI * (((float)spike + 0.5f * crandom()) / (float)spikes);
-			float altitudeVariance = 0.5f * crandom() * M_PI_2 / (float)SPIKER_MISSILEROWS;
+			float altitudeVariance = 0.5f * crandom() * M_PI_2 / (float)MISSILEROWS;
 
 			RotatePointAroundVector(dir, zenith, rowBase, RAD2DEG(azimuth));
 			RotatePointAroundVector(dir, rotAxis, dir, RAD2DEG(altitudeVariance));
```

Comment:
I think it would be good to explain somewhere that a "row" is a group of missile launch vectors with a common angle from the spiker's vertical axis (but all missiles are fired from the same origin). It took a while for me to understand this.

Reply:
I'll add such comment.

Frankly I'd rather just use some probability density function for generating the launch directions to make the code less complex, but I want the vectors to be somewhat "evenly" distributed (with large angles between them), as otherwise the damage variance goes through the roof. I don't know a better way than what I did to achieve this efficiently.

Reply:
How about [this](https://stackoverflow.com/a/26127012) algorithm? I think you could change y to be in (0, 1) instead of (-1, 1) and use it for a half sphere.

Reply:
Looks simple enough. I'll bookmark it and maybe try it out when I find the time, but I wouldn't consider the current approach with the fixes to be a PR blocker as it works reasonably well in practice.